### ISSUE TITLE
@types/react versions mismatch with upstream transient dependencies

### ISSUE THREAD

**Issue by sgoldgar** [pnpm/pnpm#5736]
Hello! We were hoping someone might be able to advise about some issues we’ve been seeing with `@types/react` in our monorepo.

**Setup**: The monorepo is on pnpm `7.14.2` and is set up with separate workspaces for several applications and packages. Some applications use various versions of NextJS and some are using various versions of React. EDIT: We are not currently using any special hoisting patterns. We are using default hoisting setup

**Issues**:
1. In the NextJS apps, there is a mis-match between the version of React when explicitly importing (i.e. `import React from ‘react’ at top of file`) vs pulling it from the global scope (i.e. just using `React.<>` in the file without import at the top). When explicitly importing, the version of `@types/react` resolves to the version defined in the application's package.json (In our case this is React 18.2.0). When using React from the global scope, the version of React defined in `next-env.d.ts`, we see that the type throwing an error is coming from an import statement in `styled-jsx/types` which is resolving to `@react/types` 17.0.45 for us.
2. We see a similar issue in our React applications that are not using NextJS where although the version of `@types/react` is correct in the application's `node_modules` folder, upstream transient dependencies are not resolving to the correct version of `@types/react` causing type definition mismatches in the downstream application.
3. We also believe that pnpm is trying to dedupe types between applications and that we might need to pin versions of `react`, `react-dom` and anything `types`. We saw that Next codegen installs with save-exact for certain dependencies - is this best practice for types?

**TL;DR** The core problem for both is that the version of @types/react isn't the same for our applications and all their upstream transient dependencies.

Any advice is much appreciated!

### pnpm version: 
7.14.2

### Expected behavior:
Different workspaces should be able to point to the version of @types/react in their package.json

### Actual behavior:
The version of @types/react isn't the same for our applications and all their upstream transient dependencies.

**Comment by zkochan** [comment]
Do you have `@react/types` in the direct dependencies? I guess you could use [`packageExtensions`](https://pnpm.io/package_json#pnpmpackageextensions) to add `@types/react` to the peer dependencies of the transient packages. In that. case the transient packages will be copied with different versions of `@types/react`.

**Comment by sgoldgar** [comment]
Thank you @zkochan for the quick response!! Just tried package extensions and it didn't work :( But we did some more investigation and it looks like our root problem might be the hard dependency link to @types/react in this 
[downstream dependency used in storybook addons](https://github.com/mdx-js/mdx/blob/main/packages/react/package.json#L45)

That is causing this in our lockfile
```
  /@mdx-js/react/2.1.5_react@18.2.0:
    resolution: {integrity: sha512-3Az1I6SAWA9R38rYjz5rXBrGKeZhq96CSSyQtqY+maPj8stBsoUH5pNcmIixuGkufYsh8F5+ka2CVPo2fycWZw==}
    peerDependencies:
      react: '>=16'
    dependencies:
      '@types/mdx': 2.0.2
      '@types/react': 18.0.26
      react: 18.2.0
    dev: false
``` 

We're about to try pnpm patch, following video that I think might have you! :)

**Comment by zkochan** [comment]
If you want to remove the dependency from that package.json, it is better to use a hook instead of a patch. You can create this [`.pnpmfile.cjs`](https://pnpm.io/pnpmfile#hooksreadpackagepkg-context-pkg--promisepkg) in the root of the workspace:

```js
module.exports = {
  hooks: {
    readPackage: (pkg) => {
      if (pkg.name === '@mdx-js/react') {
        delete pkg.dependencies['@types/react']
      }
      return pkg
    }
  }
}
```

**Comment by sgoldgar** [comment]
Hi @zkochan ! Just wanted to check back in and update that we've been able to resolve the issues short term mostly with a combination of adding `"react": ["./node_modules/@types/react"]` to the app/package `tsconfigs` and by adding `@types/react` as a peerDep using `packageExtensions`. We also noticed Vercel has what appears to be a [similar issue](https://github.com/pnpm/pnpm/issues/5631) and will follow along there as well.

**Comment by kuba-gaj** [comment]
@sgoldgar we have similar issues, could you elaborate please what exactly did you do to solve it? Thank you!

**Comment by Enrice** [comment]
you may have thought of "transitive" instead of "transient", which does have a totally different meaning?

**Comment by zkochan** [comment]
Try with pnpm v9.0.0-alpha.4

If the mismatch is between workspace projects, then you [injected dependencies](https://pnpm.io/package_json#dependenciesmetainjected)

**Comment by Brandon-Ln** [comment]
> Try with pnpm v9.0.0-alpha.4
> 
> If the mismatch is between workspace projects, then you [injected dependencies](https://pnpm.io/package_json#dependenciesmetainjected)

I don't mean to disturb. Does this kind of upstream version mismatch problem will automatically be fixed in the comming 9.X?

**Comment by milamer** [comment]
Hi there,

We encountered a similar issue and found that many packages didn't explicitly specify a dependency on `@types/react` despite relying on `react`. To address this, we implemented a solution in our `.pnpmfile.cjs` to automatically add `@types/react` to all relevant dependencies. Here's the approach we used:

```javascript
// .pnpmfile.cjs
function readPackage(pkg, context) {
  // all our workspaces start with @off
  if (pkg.name.startsWith('@off')) return pkg;

  const isReactInUse =
    'react' in pkg.dependencies ||
    'react' in pkg.devDependencies ||
    'react' in pkg.peerDependencies;
  const isTypesReactInUse =
    '@types/react' in pkg.devDependencies ||
    '@types/react' in pkg.peerDependencies ||
    '@types/react' in pkg.dependencies;

  if (isReactInUse || isTypesReactInUse) {
    delete pkg.devDependencies['@types/react'];
    const peerDependencies = pkg.peerDependencies || {};
    pkg.peerDependencies = peerDependencies;

    // our workspaces had 17 and 18 in use
    peerDependencies['@types/react'] = '17 || 18';
  }
  return pkg;
}

module.exports = {
  hooks: {
    readPackage,
  },
};
```

This modification ensures `@types/react` is set as a peer dependency with versions `17 || 18` for any package using React, accommodating our workspaces that were using different React versions. We had to use the latest 9.0 RC release of pnpm for this to work effectively.

Additionally, it would be extremely beneficial to set the correct dependency resolutions if we could access the "scope" via the context (i.e., the "path" with which a package was installed such as `["@off/web@x.x.x", "@remix-run/react@X.X.X", ...]`). This information could significantly enhance the utility and flexibility of `.pnpmfile.cjs`.

Hope this helps others facing similar issues!

**Comment by Enrice** [comment]
Same department `@types/react`:
I just had a similar issue with `pnpm@11.2.2` but not with `pnpm@10.33.4`.

When doing `pnpm -r up ...` (private, totally unrelated upstream packages without any reference to `@types/react`) I suddenly had multiple `@types/react@19.2.15` instead of `@types/react@19.2.14` as before in my `pnpm-lock.yaml`.

I checked, there is no transitive hard dependency on this explicit version.

So updating arbitrary, non-related dependencies just got worse, see #9992

**Comment by zkochan** [comment]
Closing for lack of a reproduction on a supported version.

A @types/react duplication report from 2022 on pnpm 7.14. `dedupePeerDependents` is on by default and `overrides` pins one copy; a current case needs its own reproduction.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
